apollo-federation/src/sources/connect/json_selection/methods/public/join.rs

```diff
@@ -0,0 +1,288 @@
+    method_name: &WithRange<String>,
+    method_args: Option<&MethodArgs>,
+    data: &JSON,
+    _vars: &VarsWithPathsMap,
+    input_path: &InputPath<JSON>,
+) -> (Option<JSON>, Vec<ApplyToError>) {
+    let mut warnings = vec![];
+
+    static EMPTY_STRING: String = String::new();
+    let separator = method_args
+        .and_then(|args| args.args.first())
+        .and_then(|s| match &**s {
+            LitExpr::String(s) => Some(s),
+            _ => None,
+        })
+        .unwrap_or_else(|| {
+            warnings.push(ApplyToError::new(
+                format!(
+                    "Method ->{} requires a string argument",
+                    method_name.as_ref()
+                ),
+                input_path.to_vec(),
+                method_name.range(),
+            ));
+            &EMPTY_STRING
+        });
+
+    fn to_string(value: &JSON) -> String {
+        match value {
+            JSON::Bool(b) => b.then_some("true").unwrap_or("false").to_string(),
+            JSON::Number(number) => number.to_string(),
+            JSON::String(byte_string) => byte_string.as_str().to_string(),
+            JSON::Null | JSON::Array(_) | JSON::Object(_) => "".to_string(),
+        }
+    }
+
+    let joined = match data {
+        JSON::Array(values) => values
+            .iter()
+            .map(to_string)
+            .collect::<Vec<_>>()
+            .join(separator),
+        // Single values are emitted as strings with no separator
+        _ => to_string(data),
+    };
+
+    (Some(JSON::String(joined.into())), warnings)
+}
+#[allow(dead_code)] // method type-checking disabled until we add name resolution
+fn join_method_shape(
+    method_name: &WithRange<String>,
+    method_args: Option<&MethodArgs>,
+    input_shape: Shape,
+    dollar_shape: Shape,
+    named_var_shapes: &IndexMap<&str, Shape>,
+    source_id: &SourceId,
+) -> Shape {
+    let input_shape_contract = Shape::one(
+        [
+            Shape::string([]),
+            Shape::int([]),
+            Shape::float([]),
+            Shape::bool([]),
+            Shape::null([]),
+            Shape::list(
+                Shape::one(
+                    [
+                        Shape::string([]),
+                        Shape::int([]),
+                        Shape::float([]),
+                        Shape::bool([]),
+                        Shape::null([]),
+                    ],
+                    [],
+                ),
+                [],
+            ),
+        ],
+        [],
+    );
+
+    // allow unknown input
+    if !(input_shape.is_unknown() || matches!(input_shape.case(), ShapeCase::Name(_, _))) {
+        let mismatches = input_shape_contract.validate(&input_shape);
+        if !mismatches.is_empty() {
+            return Shape::error(
+                format!(
+                    "Method ->{} requires an array of scalars values as input",
+                    method_name.as_ref()
+                ),
+                [],
+            );
```

Comment:
Can we give any idea of what the `mismatches` were with this error?

Reply:
good question! i'm not sure what to do with this. i'm pretty sure it's giving me one mismatch per shape in the One<>, so i get: 
```
expected: String, received: List<List<String>>
expected: Float, received: List<List<String>>
expected: Bool, received: List<List<String>>
expected: null, received: List<List<String>>
expected: String, received: List<String>
expected: Float, received: List<String>
expected: Bool, received: List<String>
expected: null, received: List<String>
expected: One<String, Float, Bool, null>, received: List<String>
expected: One<String, Float, Bool, null, List<One<String, Float, Bool, null>>>, received: List<List<String>>
```

that's a lot, what should we do with this?